include/irr/asset/IVirtualTexture.h

```diff
@@ -0,0 +1,750 @@
+        IVTResidentStorage(core::smart_refctd_ptr<image_t>&& _image, const core::vector<uint16_t>& _assignedLayers, uint32_t _layerShift, uint32_t _xmask) :
+            image(std::move(_image)),
+            m_assignedPageTableLayers(_assignedLayers),
+            m_addr_layerShift(_layerShift),
+            m_addr_xMask(_xmask)
+        {
+
+        }
+
+        core::smart_refctd_ptr<image_view_t> createView(E_FORMAT _format) const
+        {
+            auto found = m_viewsCache.find(_format);
+            if (found!=m_viewsCache.end())
+                return found->second;
+
+            typename image_view_t::SCreationParams params;
+            params.flags = static_cast<IImageView<image_t>::E_CREATE_FLAGS>(0);
+            params.format = _format;
+            params.subresourceRange.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+            params.subresourceRange.baseArrayLayer = 0u;
+            params.subresourceRange.layerCount = image->getCreationParameters().arrayLayers;
+            params.subresourceRange.baseMipLevel = 0u;
+            params.subresourceRange.levelCount = image->getCreationParameters().mipLevels;
+            params.image = core::smart_refctd_ptr<image_t>(image);
+            params.viewType = asset::IImageView<image_t>::ET_2D_ARRAY;
+
+            return m_viewsCache.insert({_format,createView_internal(std::move(params))}).first->second;
+        }
+        auto getPageTableLayers() const
+        {
+            return std::make_pair(m_assignedPageTableLayers.begin(),m_assignedPageTableLayers.end());
+        }
+        void addPageTableLayer(uint16_t _layer)
+        {
+            auto rng = getPageTableLayers();
+            auto it = std::lower_bound(rng.first,rng.second,_layer);
+            m_assignedPageTableLayers.insert(it, _layer);
+        }
+        void removeLayerAssignment(uint16_t _layer)
+        {
+            auto rng = getPageTableLayers();
+            auto it = std::lower_bound(rng.first, rng.second, _layer);
+            if (it!=rng.second && *it==_layer)
+                m_assignedPageTableLayers.erase(it);
+        }
+
+        inline uint32_t physPgOffset_x(SPhysPgOffset _offset) const { return _offset.addr & m_addr_xMask; }
+        inline uint32_t physPgOffset_y(SPhysPgOffset _offset) const { return (_offset.addr >> (m_addr_layerShift>>1)) & m_addr_xMask; }
+        inline uint32_t physPgOffset_layer(SPhysPgOffset _offset) const { return (_offset.addr & 0xffffu)>>m_addr_layerShift; }
+        inline bool physPgOffset_valid(SPhysPgOffset _offset) const { return (_offset.addr&0xffffu) != SPhysPgOffset::invalid_addr; }
+        inline bool physPgOffset_hasMipTailAddr(SPhysPgOffset _offset) const { return physPgOffset_valid(physPgOffset_mipTailAddr(_offset)); }
+        inline SPhysPgOffset physPgOffset_mipTailAddr(SPhysPgOffset _offset) const { return _offset.addr>>16; }
```

Comment:
these should be methods of `SPhysPgOffset`

since the bit assignments should be constexpr